llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -9164,7 +9164,8 @@ BoUpSLP::tryToGatherSingleRegisterExtractElements(
       continue;
     }
     auto *VecTy = dyn_cast<FixedVectorType>(EI->getVectorOperandType());
-    if (!VecTy || !isa<ConstantInt, UndefValue>(EI->getIndexOperand()))
+    if (!VecTy || !isa<ConstantInt, UndefValue>(EI->getIndexOperand()) ||
```

Comment:
I don't like this approach. Instead better to check if the vector operands depend on tree entires and delay the vectorization of this particular gather node. When "VectorizedValue" is ready, replace the vector operand with the vectorized value and perform shuffle.

Reply:
Dependency is actually the other way around - a gather node operand depends on a vectorized node. But the dependency is not tracked anywhere so that we end up creating a shuffle of a value which was vectorized (and is in the deleted instructions list) , I do not like the idea to introduce yet another tracking mechanism.

Reply:
Actually it does not need a new tracking mechanism, we can reuse the existing one. I can try to do it tomorrow, if you're ok with this

Reply:
Yes, I'm okay if you take it over.

Reply:
Fixed in 8ea8dd9a017182d167f39f521ef397afba5a0fd5

Reply:
Thanks!

Reply:
The problem turned out more tricky.
Slight test modification will manifest it again:
```
diff --git a/llvm/test/Transforms/SLPVectorizer/X86/gather_extract_from_vectorbuild.ll b/llvm/test/Transforms/SLPVectorizer/X86/gather_extract_from_vectorbuild.ll
index 3bccfac8566d..4fc682488d34 100644
--- a/llvm/test/Transforms/SLPVectorizer/X86/gather_extract_from_vectorbuild.ll
+++ b/llvm/test/Transforms/SLPVectorizer/X86/gather_extract_from_vectorbuild.ll
@@ -21,7 +21,7 @@ entry:

 loop:
   %ph0 = phi float [ 0.000000e+00, %entry ], [ %i4, %loop ]
-  %ph1 = phi float [ 0.000000e+00, %entry ], [ 0.000000e+00, %loop ]
+  %ph1 = phi float [ 0.000000e+00, %entry ], [ %i5, %loop ]
   %i = fadd float 0.000000e+00, %ph0
   %i1 = fadd float 0.000000e+00, %ph1
   %i2 = select i1 false, float %i, float 0.000000e+00
@@ -29,5 +29,6 @@ loop:
   %ins0 = insertelement <2 x float> zeroinitializer, float %i2, i64 0
   %ins1 = insertelement <2 x float> %ins0, float %i3, i64 1
   %i4 = extractelement <2 x float> %ins1, i64 0
+  %i5 = extractelement <2 x float> %ins1, i64 1
   br label %loop
 }
```

Reply:
Fixed in 7c963fde16d8ba340d6a6ed044b9c775f9bfab48

Reply:
Thank you, Alexey.
Still not there...
Here is another variation:
```
define void @test3() {
entry:
  br label %bb3

bb1:
  %ph = phi float [ poison, %bb2 ], [ %i5, %loop ]
  unreachable

bb2:
  br i1 poison, label %bb3, label %bb1

bb3:
  br label %loop

loop:
  %ph0 = phi float [ 0.000000e+00, %bb3 ], [ %i4, %loop ]
  %ph1 = phi float [ 0.000000e+00, %bb3 ], [ %i5, %loop ]
  %i = fadd float 0.000000e+00, %ph0
  %i1 = fadd float 0.000000e+00, %ph1
  %i2 = select i1 false, float %i, float 0.000000e+00
  %i3 = select i1 false, float %i1, float 0.000000e+00
  %ins0 = insertelement <2 x float> zeroinitializer, float %i2, i64 0
  %ins1 = insertelement <2 x float> %ins0, float %i3, i64 1
  %i4 = extractelement <2 x float> %ins1, i64 0
  %i5 = extractelement <2 x float> %ins1, i64 1
  br i1 poison, label %bb1, label %loop
}
```

Reply:
Thank you, Alexey!